### DISCUSSION TITLE
Swift, iOS 기초 공부방

### DISCUSSION THREAD

**Discussion by kyungpyoda** [boostcamp-2020/Project17-A-Map#62] [Study]
### 하나씩 떠오르는 Swift, iOS 기초 지식들을 정리해봅시닷 📖
> 참고하면 좋을..
> 제르시님의 [@JeaSungLEE/iOSInterviewquestions](https://github.com/JeaSungLEE/iOSInterviewquestions)
> 동현님의 [@TTOzzi/Question-Archive](https://github.com/TTOzzi/Question-Archive)
> [Swift Programming Language Guide](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html)
> [Swift Programming Language Guide 한글판](https://jusung.gitbook.io/the-swift-language-guide/)
> [스위프트 성능 이해하기](https://www2.slideshare.net/YongHaYoo/ss-63881606)

**Comment by kyungpyoda** [comment]
MVC 패턴에서 M, V, C 의 역할은?

**Reply by SkydevilK** [reply]
- Model
  - 특정한 객체를 캡슐화하고 해당 데이터(variable)를 조작하고 처리하는 논리와 계산(function)을 정의하는 것
  - UI와 독립되어 있다.(View와 소통이 불가능하다.)
- Controller
  - 어떻게 View에 표시할 것인지에 대한 처리
  - Model에 직접적으로 접근 가능, Model은 Notification을 통해 Model의 변화를 Controller에 알림
  - Controller는 outlet을 이용해 View에게 직접적으로 접근 가능.
- View
  - UIButton, UIViewController, UILabel와 같은 UI와 관련된 것, Controller의 통제를 받음
  - Controller에게 Action에 대한 요청(delegate)와 데이터에 대한 요청(data source)를 할 수 있음.
  - 사용자의 Action에 따라 action(View) - target(controller)의 구조로 필요한 함수를 호출 할 수 있음.

## 참고 이미지
![image](https://user-images.githubusercontent.com/48668211/101609711-a3800880-3a4a-11eb-8bcf-72b0fa21e5e7.png)

**Comment by kyungpyoda** [comment]
struct와 class, enum 타입의 차이점은?

**Reply by SkydevilK** [reply]
- class
  - 참조 타입
  - 상속 가능
  - 형 변환 가능 (상속이 되므로)
  - Deinitialization 가능 (소멸자)
  - Reference Count
- struct
  - 값 타입( 3word가 넘어 갈 경우 참조타입)
  - 구조체가 사용되는 상황
    - 연관된 간단한 값의 집합을 캡슐화하는 것만이 목적일 때
    - 캡슐화한 값을 참조하는 것보다 복사하는 것이 합당할 때 (Thread-Safe)
    - 구조체에 저장된 프로퍼티가 값 타입이며 참조하는 것보다 복사하는 것이 합당할 때
    - 다른 타입으로부터 상속받거나 자신을 상속할 필요가 없을 때
- enum
  - enum은 관련 값의 그룹에 대한 일반적인 유형을 정의
  - 각 case는 그 자체가 고유의 값
  - protocols 확장을 통해 케이스를 전부 가져올 수 있음(CaseIterable)
  - Associated Values
    - 서로 다른 타입의 값을 저장
  - Raw Values
    - 동일한 유형의 기본 값으로 미리 지정할 수 있음

>optional은 enum타입이다.
>public enum Optional<Wrapped> {
> 	case none
> 	case some(Wrapped)
>}
>따라서, 값과 없는 값을 포함하고 있고, Optional이라는 enum으로 감싸져 있는 것.

**Reply by yskpth91** [reply]
enum의 상수들은 컴파일타임에서 에러를 처리할 수 있게 도와준다. 하지만 다른 타입의 에러들은 보통 런타임에서 걸린다.(ex. 딕셔너리 ..)

**Reply by SkydevilK** [reply]
+) 값의 존재 유무에 대한 Error를 확인하는 곳이 enum은 컴파일 단계에서 알 수 있고, struct, class에서 static 상수를 만들 경우 런타임에서 확인이 가능하다.
  -> enum은 instance를 생성하지 않고 값에 접근하기 떄문

**Reply by SkydevilK** [reply]
Associated Values
서로 다른 타입의 값을 저장

효과 때문에 비교에 대한 속도가 빠르다.

**Comment by kyungpyoda** [comment]
strong과 weak의 차이점은 뭘까요? 장점과 주의해야할 단점은?

**Reply by SkydevilK** [reply]
- Strong
  - Swift에서 기본값이며 프로퍼티, 변수, 상수에 할당시 강한참조를 사용하게 된다.
  -  Reference Count가 1 증가한다.
  -  해제시엔 nil을 할당하면 Reference Count 가 다시 1 감소하게 된다.
  -  그런데 복합적으로 강한참조의 규칙을 모르고 사용하게 되면 메모리 누수(Memory Leak)가 발생

**Weak와 Unowned는 Reference Count를 증가시키지 않는게 핵심(메모리 누수 방지)**

- Weak
  - Weak은 자신이 참조하는 인스턴스의 참조 횟수를 증가시키지 않는다. 
  - 해당 객체가 nil이 할당 될 수 있는 **옵셔널 타입**이어야 한다.

- Unowned
  - Unowned는 Weak과 같은 약한 참조이지만, 객체가 nil이 할당 될 수 없는 상태 즉, **옵셔널 타입이 아니어야** 한다.
  - 인스턴스가 메모리에서 해제되지 않을때에만 사용해야 한다. 
  - Unowned 값이 nil 이라고 한다면 Crash가 발생할 수 있다.

## Strong은 프로그래머가 직접 지정해서 LifeCycle을 확실하게 지켜 줄 수 있지만, Weak은 nil로 초기화 되는 시점이 언제인지를 알 수 없다.

**Comment by kyungpyoda** [comment]
GCD Queue 종류와 동작방식 차이점?

**Reply by SkydevilK** [reply]
- Serial Queue
  - 직렬화. **한개의** Thread가 작업.
  - 작업의 순서가 보장이 됨
- Concurrent Queue
  - 동시성 
  - Queue(대기열)로 보내진 작업들을 여러개의 스레드로 보냄
  - Queue의 작업 순서가 보장이 안됨(언제 끝날지 알 수 없음)
- Main Queue
  - 오직 한개만 존재
  - Serial Queue
  - 이곳에 할당된 task는 메인 스레드에서 처리 (UI 업데이트 내용 처리)
- Global Queue
  - Concurrent Queue
  - Qos (Quality Of Service)에 따라 여러개의 종류로 나뉨
- Custom Queue
  - 커스텀으로 만듦
  - default는 Serial Queue. Concurrent로 설정 가능
  - Qos 설정 가능

**Comment by kyungpyoda** [comment]
테이블 뷰에 Custom Cell 디자인해서 화면에 출력하기까지 과정 설명

**Comment by kyungpyoda** [comment]
옵저버 패턴과 Pub-Sub 의 차이점?

**Comment by kyungpyoda** [comment]
iOS 앱 번들 구조와 샌드박스 방식

**Comment by kyungpyoda** [comment]
비동기로 방식으로 병렬처리하면 어떤 작업의 경우 성능이 개선되고, 어떤 경우는 별로 개선이 안되기도 합니다 몇가지 사례를 들어서 설명해주세요.

**Comment by kyungpyoda** [comment]
특정 메소드에 @objc를 붙이는 경우는 어떤 경우인가요? 왜 붙이는 걸까요?

**Comment by kyungpyoda** [comment]
서버에서 JSON 규격으로 데이터를 받아서 원하는 모델로 변환하기 까지 과정

**Comment by kyungpyoda** [comment]
오토레이아웃에서 특정한 상황에서 제약사항을 바꾸고 싶을때 할 수 있는 방법들을 설명해주세요.

**Comment by kyungpyoda** [comment]
뷰 객체가 그려질 때까지 제약사항, 레이아웃, 드로우 단계별로 렌더링 루프 흐름을 설명해주세요

**Comment by kyungpyoda** [comment]
코어데이와 SQLite나 다른 영구 저장소 차이점을 설명해주세요. 어떤 경우에 코어 데이터를 사용하면 좋을까요.

**Comment by hoonv** [comment]
GCD와 Operation 큐의 차이점 그 둘의 특징은 뭘까요

**Comment by SkydevilK** [comment]
왜 UI를 메인쓰레드에서 업데이트 해야할까?

**Reply by SkydevilK** [reply]
1. UIKit의 모든 속성을 Thread-safe하게 설계하면, 느려짐과 같은 성능저하가 발생할 수 있기 때문에 그렇게 설계할 수 없다. (Thread-Safe 하지 않게 설계한 것은 애플의 의도다.)
2. Main Thread가 View의 업데이트를 관리하는 View Drawing Cycle을 통해 View를 동시에 업데이트 하는 그런 설계를 통해 동작하고 있는데, (Main Thread가 아닌) BackGround Thread가 각자의 Run-Loop로 그런 동작을 하게되었을때, View가 제멋대로 동작할 수있다. (예를  들어, 기기를 회전 했을때, 동시에 View의 Layout이 재배치되는 그런 동작을 못하게 될 수도 있다.)
3. iOS가 그림의 그리는 Rendering Process(CoreAnimation -> Render Server -> GPU -> Present)가 있는데, 여러 Thread에서 각자의 View의 변경사항 GPU로 보내면 GPU는 각각의 정보를 다 해석해야하니 느려지거나, 비효율적이 될 수 있다.
4. Texture나 ComponentKit이라는 페이스북에서 개발한 비동기적 UI Framework가 있긴 하지만, View Drawing Cycle가 유사한 방식으로 적절한 타이밍에 Main Thread에서 동시에 업데이트 하도록 하고 있다.
